File: icon-designer/backend/src/main/kotlin/de/ilijaz/myapp/core/util/Preconditions.kt
package de.ilijaz.myapp.core.util

class Preconditions(override val message: String) : RuntimeException(message) {
    companion object {
        fun <T> notNull(obj: T?): T = obj ?: throw Preconditions("not null assertion exception")

        inline fun <reified T> type(obj: Any) {
            if (obj !is T) {
                throw Preconditions("not of type ${T::class} assertion exception")
            }
        }
    }
}

File: icon-designer/frontend/src/core/preconditions.ts
export const Preconditions = {
  notNull<T>(obj: T | null | undefined): T {
    if (obj == null) {
      throw new Error('not null assertion error');
    }
    return obj;
  },
  in<T>(obj: T, list: T[]): T {
    if (!list.includes(obj)) {
      throw new Error('in assertion error');
    }
    return obj;
  },
};
